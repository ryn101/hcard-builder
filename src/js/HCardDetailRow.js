import Unique from './Unique';
import HCardDetail from './HCardDetail';

/**
 * Class storing details specific to a hCard detail row component
 */
class HCardDetailRow extends Unique {
    constructor() {
        super('hcdr');
        this.details = [];
    }

    /**
     * Adds a single detail to the HCardDetailRow component
     *
     * @param label The label for the detail
     * @param properties The properties to be formatted as a single line
     * @param propertiesSeparator The separator for the each hCard property
     * @returns {HCardDetailRow} Returns the HCardDetailRow instance
     */
    addDetail(label, properties, propertiesSeparator) {
        this.details.push(new HCardDetail(label, properties, propertiesSeparator));

        return this;
    }
}

export default HCardDetailRow;
