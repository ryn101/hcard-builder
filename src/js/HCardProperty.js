/**
 * Configured types of hCard properties used as a lookup for value safety
 *
 * @type {{
 *    country: string,
 *    streetAddress: string,
 *    postalCode: string,
 *    locality: string,
 *    telephone: string,
 *    region: string,
 *    email: string
 *  }}
 */
const hCardPropertyType = {
    email: 'email',
    telephone: 'tel',
    streetAddress: 'street-address',
    locality: 'locality',
    region: 'region',
    postalCode: 'postal-code',
    country: 'country',
};

/**
 * Container for a HCard specific property
 */
class HCardProperty {
    /**
     * Constructor
     *
     * @param name The property name/type
     * @param value The value of the property
     */
    constructor(name, value) {
        this.name = name;
        this.value = value;
    }
}

export { hCardPropertyType as HCardPropertyType };
export default HCardProperty;
