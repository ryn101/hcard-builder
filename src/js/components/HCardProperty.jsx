import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

/**
 * Wraps up the logic needed to display a hCard property. The name of the property is added
 * as a class to the component and the title is set to the specified value.
 *
 * Additionally supports property nesting, and abbreviation declaration for use with the abbr tag.
 * If the abbreviation tag is provided then the as property is disregarded.
 * If child elements are provided then the value and abbreviation properties are disregarded.
 *
 * When requiring a HTML tag other than a span, the as property can be specified.
 */
function HCardProperty(props) {
    const {
        name,
        value,
        abbreviation,
        as,
        className,
        children,
    } = props;

    const ElementTag = as;
    let element;

    if (children || !abbreviation) {
        // If children are specified or no abbreviation value is specified then use the element tag
        element = (
            <ElementTag title={value} className={classNames(name, className)}>
                { children || value }
            </ElementTag>
        );
    } else {
        // If an abbreviation is specified then create an abbr tag instead
        element = (<abbr title={value} className={name}>{abbreviation}</abbr>);
    }

    return element;
}

HCardProperty.propTypes = {
    name: PropTypes.string.isRequired,
    value: PropTypes.string,
    abbreviation: PropTypes.string,
    as: PropTypes.string,
    className: PropTypes.string,
    children: PropTypes.oneOfType([
        PropTypes.element,
        PropTypes.string,
        PropTypes.arrayOf(PropTypes.oneOfType([
            PropTypes.element,
            PropTypes.string,
        ])),
    ]),
};

HCardProperty.defaultProps = {
    value: '',
    as: 'span',
    className: '',
    children: null,
};

export default HCardProperty;
