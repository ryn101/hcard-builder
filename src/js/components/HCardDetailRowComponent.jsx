import React from 'react';
import PropTypes from 'prop-types';
import HCardDetailRow from '../HCardDetailRow';
import HCardPropertyComponent from './HCardPropertyComponent';

const HCardDetailRowComponent = ({ hCardDetailRow }) => (
    <li className={`flex-row flex-row--columns-${hCardDetailRow.details.length}`}>
        {
            hCardDetailRow.details.map((detail) => (
                <div className="detail flex-column" key={detail.id}>
                    <div>
                        <span className="label">{ detail.label }</span>
                        <span className="value">
                            {
                                detail.properties.filter((property) => property.value)
                                    .map((property) => (
                                        <HCardPropertyComponent
                                            key={property.name}
                                            hCardProperty={property}
                                        />
                                    )).flatMap((value, index, array) => (
                                        // Add in the separator element
                                        array.length - 1 !== index
                                            ? [value, detail.propertiesSeparator]
                                            : value
                                    ))
                            }
                        </span>
                    </div>
                </div>
            ))
        }
    </li>
);

HCardDetailRowComponent.propTypes = {
    hCardDetailRow: PropTypes.instanceOf(HCardDetailRow).isRequired,
};

export default HCardDetailRowComponent;
