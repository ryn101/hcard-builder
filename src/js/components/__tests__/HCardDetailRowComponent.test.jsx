import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import HCardDetailRowComponent from '../HCardDetailRowComponent';
import HCardDetailRow from '../../HCardDetailRow';
import HCardProperty from '../../HCardProperty';

configure({ adapter: new Adapter() });

beforeAll(() => {
    HCardDetailRow.resetCounters();
});

it('renders the hCard detail row without crashing', () => {
    const hCardDetailRow = new HCardDetailRow();
    const div = document.createElement('div');
    ReactDOM.render(<HCardDetailRowComponent hCardDetailRow={hCardDetailRow} />, div);
    ReactDOM.unmountComponentAtNode(div);
});

describe('once the hcard detail row has rendered', () => {
    let hCardDetailRowNode;
    const hCardDetailRow = new HCardDetailRow()
        .addDetail('Detail 1', [new HCardProperty('Prop1', 'Value1')]);

    beforeEach(() => {
        hCardDetailRowNode = shallow(<HCardDetailRowComponent hCardDetailRow={hCardDetailRow} />);
    });

    it('renders with state correctly', () => {
        expect(hCardDetailRowNode).toMatchSnapshot();
    });

    it('renders the required details', () => {
        const detailNodes = hCardDetailRowNode.find('.detail');
        expect(detailNodes).toHaveLength(hCardDetailRow.details.length);

        const detailNode = detailNodes.first();
        const detail = hCardDetailRow.details[0];

        expect(detailNode.find('.label').text()).toEqual(detail.label);

        const valueChildNodes = detailNode.find('.value').children();
        expect(valueChildNodes).toHaveLength(detail.properties.length);
    });
});
