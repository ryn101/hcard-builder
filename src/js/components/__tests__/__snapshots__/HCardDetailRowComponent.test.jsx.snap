// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`once the hcard detail row has rendered renders with state correctly 1`] = `
<li
  className="flex-row flex-row--columns-1"
>
  <div
    className="detail flex-column"
    key="hcd-1"
  >
    <div>
      <span
        className="label"
      >
        Detail 1
      </span>
      <span
        className="value"
      >
        <HCardPropertyComponent
          className="Prop1"
          hCardProperty={
            HCardProperty {
              "name": "Prop1",
              "value": "Value1",
            }
          }
          key="Prop1"
          title="Value1"
        />
      </span>
    </div>
  </div>
</li>
`;
